### PR TITLE
new-app: appropriately warn/error on circular builds

### PR DESCRIPTION
**Pull request by gabemontero**
Bug 1393549
Bugzilla link https://bugzilla.redhat.com/show_bug.cgi?id=1393549
Detailed explanation
Citing of cirular builds were performed for new-build but not new-app;
With new-app, if template, simply warn, if not template fail command

@csrwng PTAL

### CODE REVIEW

File: pkg/generate/app/cmd/newapp.go [hunk 1]
```diff
@@ -710,20 +710,29 @@ func (c *AppConfig) Run() (*AppResult, error) {
...
 				if len(c.To) == 0 {
 					// Output reference was generated, return error.
 					return nil, fmt.Errorf("%v, set a different tag with --to", err)
 				}
 				// Output reference was explicitly provided, print warning.
 				fmt.Fprintf(c.ErrOut, "--> WARNING: %v\n", err)
 			} else {
-				return nil, err
+				// circular reference handling for `oc new-app`
+				if len(templateObjects) > 0 {
```
Comment (csrwng) [inline pkg/generate/app/cmd/newapp.go]: @gabemontero it is possible for us to be generating new objects as well as processing template objects. We should check for circular references on the 2 sets of objects separately if we want different behaviors for each.

Comment (gabemontero) [inline pkg/generate/app/cmd/newapp.go]: Interesting ... so only allow the circular refs if they were induced by the template objects specifically.  OK, I'll work on a refactor.

File: test/testdata/circular-is.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (stevekuznetsov) [inline test/testdata/circular-is.json]: The word I've heard from on high is to just use YAML everywhere. If we want examples of JSON, we can have them in actual examples (under `examples/...`) but for stuff devs are working with the decision was YAML.

Comment (gabemontero) [inline test/testdata/circular-is.json]: ok i'll convert

**Comment by gabemontero** [comment]
updates pushed ... ptal

File: test/testdata/circular-is.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+items:
+- apiVersion: v1
+  kind: ImageStream
+  metadata:
+    creationTimestamp: 2016-12-02T18:55:30Z
```
Comment (csrwng) [inline test/testdata/circular-is.yaml]: you should likely get rid of these fields (creationTimestamp, generation, resourceVersion, selfLink, uid)

File: test/testdata/circular-is.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    generation: 1
+    name: newapp-circular-template
+    resourceVersion: "764"
+    selfLink: /oapi/v1/namespaces/myproject/imagestreams/newapp-circular-template
+    uid: e5ad1d91-b8c0-11e6-9803-68f7287398a7
+  spec: {}
+  status:
```
Comment (csrwng) [inline test/testdata/circular-is.yaml]: Remove the status part

File: test/testdata/circular.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
+apiVersion: v1
+kind: Template
+metadata:
+  creationTimestamp: null
```
Comment (csrwng) [inline test/testdata/circular.yaml]: get rid of creationTimestamp in this file

**Comment by gabemontero** [comment]
yaml trimmed

**Comment by csrwng** [comment]
LGTM [merge]

**Comment by gabemontero** [comment]
flake https://github.com/openshift/origin/issues/12112

@csrwng please repost the merge comment at your convenience, and we'll try again

**Comment by csrwng** [comment]
#12112
[merge]

**Pull request merged**